### PR TITLE
fix: simpGoal reports incomplete UsedSimps

### LINKED ISSUE #2486: simpGoal reports incomplete UsedSimps
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover/lean4/issues).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

`simpGoal` does not pass the `UsedSimps` from the hypotheses to the method that simplifies the target. As a result, `simp? at *` does not report simp theorems that were only used to simplify the hypotheses.

### Steps to Reproduce

Sort of minimal working example (based on the `simp_trace` test):

```lean
def h (x : Nat) : Sum (Nat × Nat) Nat := Sum.inl (x, x)

def bla (x : Nat) :=
  match h x with
  | Sum.inl (y, z) => y + z
  | Sum.inr _ => 0

example (h' : bla x = x) : x + x = x := by
  set_option tactic.simp.trace true in
  simp [bla, h] at *
  -- Expected: Try this: simp only [bla, h] at *
  -- Actual:   Try this: simp only at *
  exact h'
```

### Versions

Current master.

### PR DESCRIPTION
**Pull request by JLimperg**
* [X] Put an X in this bracket to confirm you have read the
  [External Contribution Guidelines](https://github.com/leanprover/lean4/blob/master/doc/contributions.md).

Fixes #2486

### CODE REVIEW

**Comment by semorrison** [comment]
I'd earlier made #2333, but didn't finish writing tests. I'll close that in favour of this.

File: tests/lean/simp_trace.lean [hunk 1]
```diff
@@ -81,3 +81,7 @@ example : a ∧ (b ∧ b) ↔ a ∧ b := by simp [my_thm]
 example : (a ∧ (b ∧ b)) = (a ∧ b) := by simp only [my_thm]
 
 example : x - 1 + 1 = x := by simp (discharger := sorry) [Nat.sub_add_cancel]
+
+example (h' : bla x = x) : x + x = x := by
+  simp [bla, h] at *
+  exact h'
```
Comment (semorrison) [inline tests/lean/simp_trace.lean:87]: To be thorough, could we also have the tests here at use `at h'`, and another that uses `at *` in which there are actually multiple hypotheses that get simplified, and different simp lemmas are used at each hypothesis?

Comment (JLimperg) [reply tests/lean/simp_trace.lean:87]: Done. Thanks for the review!

**Comment by JLimperg** [comment]
awaiting-review

**Comment by semorrison** [comment]
Two more requests for tests:
* a `simp at *` where a lemma is used at two hypotheses, and check it is only reported once
* similarly where is used at a hypothesis and a goal
* perhaps already covered, but one where a provided simp lemma in the `[...]` arguments is not used at all.

**Comment by JLimperg** [comment]
Also done. Btw, I believe we could use your mass-testing code to check whether there are any remaining issues. I have code that can check whether two tactics generate the same goals, so we could run every `simp` invocation in mathlib as `simp?` and check whether the generated `Try this` suggestion generates the same goals.

**Comment by JLimperg** [comment]
awaiting-review

**Review by semorrison — APPROVED** [review]

**Comment by tydeu** [comment]
@semorrison The mathlib bot appears to have built this. I presume the absence of a bot commit on the status is due to this occur between changes to the CI?

**Comment by semorrison** [comment]
Yes, the "send the result back to the Lean4 PR" commit was only merged to Mathlib master after this.

**Comment by semorrison** [comment]
I had not realised that the "added a commit" message was going to be automatically generated (because I put the URL for this PR in the commit message!). It occurs to me now that this may be sufficient feedback, and there is no need to write comments or use labels. Let's see how this pans out over the next few days.

**Comment by tydeu** [comment]
@semorrison 

> It occurs to me now that this may be sufficient feedback, and there is no need to write comments or use labels.

The labels are still useful regardless when looking through the issues on the issue list. While the comment is less useful,, it does importantly produce a notification after the build succeeds or fails whereas the commit reference produces none.

**Pull request merged**